rust/lance-encoding/src/encodings/physical/binary.rs

```diff
@@ -457,6 +462,220 @@ fn get_indices_from_string_arrays(
+
+                let this_chunk_size = (num_values_in_this_chunk + 1) * 4
+                    + (offsets[offsets.len() - 1] - offsets[last_offset_in_orig_idx]) as usize;
+
+                let padded_chunk_size = ((this_chunk_size + 7) / 8) * 8;
+
+                // the bytes are put after the offsets
+                let this_chunk_bytes_start_offset = (num_values_in_this_chunk + 1) * 4;
+                chunks_info.push(ChunkInfo {
+                    chunk_start_offset_in_orig_idx: last_offset_in_orig_idx,
+                    chunk_last_offset_in_orig_idx: this_last_offset_in_orig_idx,
+                    bytes_start_offset: this_chunk_bytes_start_offset,
+                    padded_chunk_size,
+                });
+                chunks.push(MiniBlockChunk {
+                    log_num_values: 0,
+                    num_bytes: padded_chunk_size as u16,
+                });
+                break;
+            } else {
+                // case 2: not the last chunk
+                let num_values_in_this_chunk =
+                    this_last_offset_in_orig_idx - last_offset_in_orig_idx;
+
+                let this_chunk_size = (num_values_in_this_chunk + 1) * 4
+                    + (offsets[this_last_offset_in_orig_idx] - offsets[last_offset_in_orig_idx])
+                        as usize;
+
+                let padded_chunk_size = ((this_chunk_size + 7) / 8) * 8;
+
+                // the bytes are put after the offsets
+                let this_chunk_bytes_start_offset = (num_values_in_this_chunk + 1) * 4;
+
+                chunks_info.push(ChunkInfo {
+                    chunk_start_offset_in_orig_idx: last_offset_in_orig_idx,
+                    chunk_last_offset_in_orig_idx: this_last_offset_in_orig_idx,
+                    bytes_start_offset: this_chunk_bytes_start_offset,
+                    padded_chunk_size,
+                });
+
+                chunks.push(MiniBlockChunk {
+                    log_num_values: num_values_in_this_chunk.trailing_zeros() as u8,
+                    num_bytes: padded_chunk_size as u16,
+                });
+
+                last_offset_in_orig_idx = this_last_offset_in_orig_idx;
+            }
+        }
+        let output_total_bytes = chunks_info
+            .iter()
+            .map(|chunk_info| chunk_info.padded_chunk_size)
+            .sum::<usize>();
+
+        let mut output: Vec<u8> = Vec::with_capacity(output_total_bytes);
+        for chunk in chunks_info {
+            // `this_chunk_offsets` are offsets that points to bytes in this chunk,
+            let this_chunk_offsets = offsets
+                [chunk.chunk_start_offset_in_orig_idx..chunk.chunk_last_offset_in_orig_idx + 1]
+                .iter()
+                .map(|offset| {
+                    offset - offsets[chunk.chunk_start_offset_in_orig_idx]
+                        + chunk.bytes_start_offset as u32
+                })
+                .collect::<Vec<_>>();
```

Comment:
It appears you are copying `n+1` offsets and normalizing them so they always start at 0.  If you're going to normalize, there is no need to copy the 0.

Alternatively, you could skip the normalization, and copy the n+1 values.  The decompressor would then need to subtract the first value from everything.

Not sure which approach is better, and this approach is fine for now, but it is technically doing more work than needed.

Reply:
> It appears you are copying n+1 offsets and normalizing them so they always start at 0.

they normalize to `chunk.bytes_start_offset`.
I can skip the normalization here but then I also need to store the `chunk.bytes_start_offset` in the `output` buffer so the `decompressor` know the `chunk.bytes_start_offset`